Sources/StreamChat/Repositories/Operations/SyncOperations.swift

```diff
@@ -0,0 +1,140 @@
+                let cids = database.backgroundReadOnlyContext.loadAllChannelListQueries()
+                    .flatMap(\.channels)
+                    .compactMap { try? ChannelId(cid: $0.cid) }
+                log.info("Retrieved channels from existing queries from DB. Count \(cids.count)", subsystems: .offlineSupport)
+                context.localChannelIds = cids
+                done(.continue)
+            }
+        }
+    }
+}
+
+class GetPendingConnectionDateOperation: AsyncOperation {
+    init(database: DatabaseContainer, context: SyncContext) {
+        super.init(maxRetries: 2) { [weak database] done in
+            database?.backgroundReadOnlyContext.perform {
+                context.lastPendingConnectionDate = database?.backgroundReadOnlyContext.currentUser?.lastPendingConnectionDate
+                done(.continue)
+            }
+        }
+    }
+}
+
+class SyncEventsOperation: AsyncOperation {
+    init(database: DatabaseContainer, syncRepository: SyncRepository, context: SyncContext) {
+        super.init(maxRetries: 2) { [weak database, weak syncRepository] done in
+            log.info(
+                "1. Call `/sync` endpoint and get missing events for all locally existed channels",
+                subsystems: .offlineSupport
+            )
+            guard let lastPendingConnectionDate = context.lastPendingConnectionDate else {
+                done(.continue)
+                return
+            }
+
+            syncRepository?.syncMissingEvents(
+                using: lastPendingConnectionDate,
+                channelIds: context.localChannelIds,
+                bumpLastSync: false
+            ) { result in
+                switch result {
+                case let .success(channelIds):
+                    context.synchedChannelIds = Set(channelIds)
+                    // As per our sync logic, we should keep the last connection date.
+                    database?.write { session in
+                        session.currentUser?.lastPendingConnectionDate = context.lastConnectionDate
+                    } completion: { _ in done(.continue) }
+                case let .failure(error):
+                    context.synchedChannelIds = Set([])
+                    done(error.shouldRetry ? .retry : .continue)
+                }
+            }
+        }
+    }
+}
+
+class WatchChannelOperation: AsyncOperation {
+    init(controller: ChatChannelController, context: SyncContext) {
+        super.init(maxRetries: 2) { [weak controller] done in
+            guard let controller = controller, controller.isAvailableOnRemote else {
+                done(.continue)
+                return
+            }
+
+            // Reset only if it needs recovery
+            if let cid = controller.cid, context.synchedChannelIds.contains(cid) {
+                done(.continue)
+                return
+            }
+
+            let cidString = (controller.cid?.rawValue ?? "unknown")
```

Comment:
Do we want to fall back to "unknown" here or is a missing cid worthy of an error?

Reply:
I guess it will fall through but still make a query with unknown?

Reply:
A cid is not needed here (AKA it might not be there yet), this `cidString` is only for debug purposes